staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/conversion_test.go

```diff
@@ -18,13 +18,395 @@ package v1
 
 import (
 	"reflect"
+	"strings"
 	"testing"
 
-	"k8s.io/apimachinery/pkg/runtime"
+	"github.com/google/go-cmp/cmp"
 
 	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/utils/pointer"
 )
 
+func TestConversion(t *testing.T) {
+	testcases := []struct {
+		Name      string
+		In        runtime.Object
+		Out       runtime.Object
+		ExpectOut runtime.Object
+		ExpectErr string
+	}{
+		// Versions
+		{
+			Name:      "internal to v1, no versions",
+			In:        &apiextensions.CustomResourceDefinition{},
+			Out:       &CustomResourceDefinition{},
+			ExpectOut: &CustomResourceDefinition{},
+		},
+		{
+			Name: "internal to v1, top-level version",
+			In: &apiextensions.CustomResourceDefinition{
+				Spec: apiextensions.CustomResourceDefinitionSpec{
+					Version: "v1",
+				},
+			},
+			Out: &CustomResourceDefinition{},
+			ExpectOut: &CustomResourceDefinition{
+				Spec: CustomResourceDefinitionSpec{
+					Versions: []CustomResourceDefinitionVersion{{Name: "v1", Served: true, Storage: true}},
+				},
+			},
+		},
+		{
+			Name: "internal to v1, multiple versions",
```

Comment:
I assume there's an identical variant of this for v1beta1?  i.e. units guard this already in other versions?

Reply:
internal and v1beta1 types match fields, so the conversion is just auto-generated.

v1beta1 defaulting ensures that empty version/versions are defaulted:
https://github.com/kubernetes/kubernetes/blob/32dc42ed34eb5ca97bd8fa4544d9a91d6b05a310/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/defaults.go#L52-L63

validation ensures that spec.version matches spec.versions[0].name:
https://github.com/kubernetes/kubernetes/blob/32dc42ed34eb5ca97bd8fa4544d9a91d6b05a310/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L236-L238

and that spec.versions are specified:
https://github.com/kubernetes/kubernetes/blob/32dc42ed34eb5ca97bd8fa4544d9a91d6b05a310/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L185-L191

https://github.com/kubernetes/kubernetes/blob/32dc42ed34eb5ca97bd8fa4544d9a91d6b05a310/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L229-L231

Reply:
and there's extensive validation and defaulting units for v1beta1